shared/src/output.rs

```diff
@@ -0,0 +1,365 @@
+    sections.append(&mut body.sections);
+
+    let text = BuildpackOutputText {
+        default_code: Some(String::from(ANSI_YELLOW_CODE)),
+        line_prefix: Some(String::from("! ")),
+        sections,
+        ..BuildpackOutputText::default()
+    };
+
+    println!("{}", text.to_ansi_string());
+}
+
+pub fn print_error(title: impl AsRef<str>, body: impl Into<BuildpackOutputText>) {
+    let title = title.as_ref();
+
+    let mut sections = vec![BuildpackOutputTextSection::regular(format!(
+        "ERROR: {title}\n\n"
+    ))];
+
+    let mut body = body.into();
+    sections.append(&mut body.sections);
+
+    let text = BuildpackOutputText {
+        default_code: Some(String::from(ANSI_RED_CODE)),
+        line_prefix: Some(String::from(ERROR_WARNING_LINE_PREFIX)),
+        sections,
+        ..BuildpackOutputText::default()
+    };
+
+    println!("{}", text.to_ansi_string());
+}
+
+pub fn run_command<E, F: FnOnce(std::io::Error) -> E, F2: FnOnce(Output) -> E>(
+    mut command: Command,
+    quiet: bool,
+    io_error_fn: F,
+    exit_status_fn: F2,
+) -> Result<Output, E> {
+    let child = if quiet {
+        command.output_and_write_streams(std::io::sink(), std::io::sink())
+    } else {
+        command.output_and_write_streams(
+            line_mapped(std::io::stdout(), add_prefix_to_non_empty("    ")),
+            line_mapped(std::io::stderr(), add_prefix_to_non_empty("    ")),
+        )
+    };
+
+    child.map_err(io_error_fn).and_then(|output| {
+        if output.status.success() {
+            Ok(output)
+        } else {
+            Err(exit_status_fn(output))
+        }
+    })
+}
+
+fn add_prefix_to_non_empty<P: Into<Vec<u8>>>(prefix: P) -> impl Fn(Vec<u8>) -> Vec<u8> {
+    let prefix = prefix.into();
+
+    move |mut input| {
+        if input.is_empty() {
+            vec![]
+        } else {
+            let mut result = prefix.clone();
+            result.append(&mut input);
+            result
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct BuildpackOutputText {
+    pub line_prefix: Option<String>,
+    pub default_code: Option<String>,
+    pub reset_code: String,
+    pub value_code: String,
+    pub sections: Vec<BuildpackOutputTextSection>,
+}
+
+impl Default for BuildpackOutputText {
+    fn default() -> Self {
+        Self {
+            line_prefix: None,
+            default_code: None,
+            reset_code: String::from(ANSI_RESET_CODE),
+            value_code: String::from(ANSI_VALUE_CODE),
+            sections: vec![],
+        }
+    }
+}
+
+impl BuildpackOutputText {
+    pub fn new(sections: impl Into<Vec<BuildpackOutputTextSection>>) -> Self {
+        Self {
+            sections: sections.into(),
+            ..Self::default()
+        }
+    }
+
+    fn to_ansi_string(&self) -> String {
+        let mut result = String::new();
+
+        // Every line must start with a style reset, the default ANSI code and the line prefix if
+        // it exists.
```

Comment:
Needs to be the other way around. Each line must end with a reset code, but resume the prior code on the next line.

Reply:
I think this is for handling colorizing one value while inside of another? If so, maybe we update the comment to say why that's the case.